prov/verbs/src/verbs_domain_xrc.c

```diff
@@ -0,0 +1,576 @@
+	 * avoid the appearance of stale/duplicate CM messages */
+	if (!ep->id->qp)
+		conn_param.qp_num = ep->conn_setup->rsvd_ini_qpn->qp_num;
+
+	assert(ep->conn_state == FI_IBV_XRC_UNCONNECTED ||
+	       ep->conn_state == FI_IBV_XRC_ORIG_CONNECTED);
+	ep->conn_state++;
+
+	ret = rdma_connect(ep->id, &conn_param) ? -errno : 0;
+	if (ret) {
+		ret = -errno;
+		VERBS_INFO_ERRNO(FI_LOG_FABRIC, "rdma_connect", errno);
+		ep->conn_state--;
+	}
+	return ret;
+}
+
+int fi_ibv_ep_create_tgt_qp(struct fi_ibv_ep *ep, uint32_t tgt_qpn)
+{
+	struct ibv_qp_open_attr open_attr;
+	struct ibv_qp_init_attr_ex attr_ex;
+	struct fi_ibv_domain *domain = fi_ibv_msg_ep_to_domain(ep);
+	struct ibv_qp *rsvd_qpn;
+
+	assert(ep->tgt_id && !ep->tgt_id->qp);
+
+	/* If a target QP number was specified then open that existing
+	 * QP for sharing. */
+	if (tgt_qpn) {
+		rsvd_qpn = fi_ibv_reserve_qpn(ep);
+		if (!rsvd_qpn) {
+			VERBS_WARN(FI_LOG_FABRIC,
+				   "Create of XRC reserved QPN failed %d\n",
+				   errno);
+			return -errno;
+		}
+
+		memset(&open_attr, 0, sizeof(open_attr));
+		open_attr.qp_num = tgt_qpn;
+		open_attr.comp_mask = IBV_QP_OPEN_ATTR_NUM |
+			IBV_QP_OPEN_ATTR_XRCD | IBV_QP_OPEN_ATTR_TYPE |
+			IBV_QP_OPEN_ATTR_CONTEXT;
+		open_attr.xrcd = domain->xrc.xrcd;
+		open_attr.qp_type = IBV_QPT_XRC_RECV;
+		open_attr.qp_context = ep;
+
+		ep->tgt_ibv_qp = ibv_open_qp(domain->verbs, &open_attr);
+		if (!ep->tgt_ibv_qp) {
+			VERBS_INFO_ERRNO(FI_LOG_EP_CTRL,
+				   "XRC TGT QP, ibv_open_qp()", errno);
+			fi_ibv_release_qpn(rsvd_qpn);
+			return -errno;
+		}
+		ep->conn_setup->rsvd_tgt_qpn = rsvd_qpn;
+		return FI_SUCCESS;
+	}
+
+	/* An existing XRC target was not specified, create XRC TGT
+	 * side of new physical connection. */
+	fi_ibv_msg_ep_get_qp_attr(ep, NULL,
+			(struct ibv_qp_init_attr *)&attr_ex);
+	attr_ex.qp_type = IBV_QPT_XRC_RECV;
+	attr_ex.qp_context = ep;
+	attr_ex.comp_mask = IBV_QP_INIT_ATTR_PD | IBV_QP_INIT_ATTR_XRCD;
+	attr_ex.pd = domain->pd;
+	attr_ex.xrcd = domain->xrc.xrcd;
+	if (rdma_create_qp_ex(ep->tgt_id, &attr_ex)) {
+		VERBS_INFO_ERRNO(FI_LOG_EP_CTRL,
+				 "Physical XRC TGT QP, rdma_create_ep_ex()",
+				 errno);
+		return -errno;
+	}
+	ep->tgt_ibv_qp = ep->tgt_id->qp;
+
+	return FI_SUCCESS;
+}
+
+static int fi_ibv_put_tgt_qp(struct fi_ibv_ep *ep)
+{
+	int ret;
+
+	if (!ep->tgt_ibv_qp)
+		return FI_SUCCESS;
+
+	/* The kernel will not destroy the detached TGT QP until all
+	 * shared opens have called ibv_destroy_qp. */
+	ret = ibv_destroy_qp(ep->tgt_ibv_qp);
+	if (ret) {
+		VERBS_INFO_ERRNO(FI_LOG_EP_CTRL,
+				 "Close XRC TGT QP, ibv_destroy_qp()", errno);
+		return -errno;
+	}
+	ep->tgt_ibv_qp = NULL;
+
+	return FI_SUCCESS;
+}
+
+int fi_ibv_ep_destroy_xrc_qp(struct fi_ibv_ep *ep)
+{
+	struct fi_ibv_domain *domain = fi_ibv_msg_ep_to_domain(ep);
+
+	if (ep->ibv_qp) {
+		fastlock_acquire(&domain->xrc.ini_mgmt_lock);
+		fi_ibv_put_shared_ini_conn(ep);
+		fastlock_release(&domain->xrc.ini_mgmt_lock);
```

Comment:
We're in a destroy path, is a lock needed here?

Reply:
I thought so. The ini_conn object is shared between other endpoints and owned by the domain. It protects the RBTree and the entry. Is there is a guarantee that another endpoint will not have a connection outstanding, or initiate a connection?